File: DynamicProgramming/coinProblem.py
# -*- coding: utf-8 -*-
"""
Created on Thu Aug  6 18:29:13 2020

@author: Goutam Dadhich
"""

import sys

INT_MAX = sys.maxsize
# print(INT_MAX)

# Finding Minimum Coins using Recursio
def FindMinCoins_rec(coins, n, amt):
    if amt == 0:
        return 0
    
    if amt < 0:
        return INT_MAX
    
    count = INT_MAX
    
    for i in range(0, n):
        res = FindMinCoins_rec(coins, n, amt-coins[i])
        
        if res != INT_MAX:
            count = min(res+1, count)
            
    return count

# Finding Minimum Coins using Dynamic programming
def FindMinCoins_dp(coins, n, amt):
    T = [INT_MAX]*(amt+1)
    
    T[0] = 0
    
    for i in range(1, amt+1):
        res = INT_MAX
        
        for j in range(0, n):
            if i-coins[j] >= 0:
                res = T[i - coins[j]]
                
            if res != INT_MAX:
                T[i] = min(res+1, T[i])
                
    return T[amt]


if __name__ == '__main__':
    coins = [1, 2, 3, 5, 7]
    n = len(coins)
    amt = 15

    print('-'*10 + '*'*5 + '-'*10)
    print('Using Recurssion:- ')
    print('Minimum coins required :- ', FindMinCoins_rec(coins, n, amt))
    print('-'*10 + '*'*5 + '-'*10)
    print('Using Recurssion:- ')
    print('Minimum coins required :- ', FindMinCoins_dp(coins, n, amt))
    print('-'*10 + '*'*5 + '-'*10)
    

File: DynamicProgramming/coinProblem.cpp
#include<stdio.h>
#include<limits.h>

int min(int x, int y)
{
	return x<y?x:y;
}

int FindMin_Recurssion(int *C, int n, int amt)
{
	if(amt == 0)
	return 0;
	
	if(amt < 0)
	return INT_MAX;
	
	int count, res;
	
	count = INT_MAX;
	
	for(int i=0; i<n; i++)
	{
		res = FindMin_Recurssion(C, n, amt-C[i]);
		
		if (res != INT_MAX)
		count = min(count, res+1);
	}
	
	
	
	return count;
}


int FindMin_DP(int *C, int n, int amt)
{
	int arr[amt+1];
	arr[0] = 0;
	
	for(int i = 1; i<=amt; i++)
	{
		arr[i] = INT_MAX;
		int res = INT_MAX;
		
		for(int j = 0; j<n; j++)
		{
			if (i - C[j] >= 0)
			res = arr[i - C[j]];
			
			if (res != INT_MAX)
			arr[i] = min(arr[i], res+1);
				
		}
	}
	
	return arr[amt];
}


int main()
{
    int C[] = {2,3,5,7};
    int amt = 15;
    int n;
    n = sizeof(C)/sizeof(C[0]);
    
    printf("\n------------********---------------\n");
    
    printf("Using recurssion :- \n");
    int res_r = FindMin_Recurssion(C, n, amt);
    printf("Minimum coins required :   %d", res_r);
    
    printf("\n------------********---------------\n");
    
    printf("Using dynamic programming :- \n");
    int res_dp = FindMin_DP(C, n, amt);
    printf("Minimum coins required :   %d", res_dp);
    
    printf("\n------------********---------------\n");
	    
    return 0;
}
